From 4d5ad77939863aef4447973b3f26847f573f83ab Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Wed, 17 Jan 2018 15:53:53 +0100
Subject: [PATCH 143/261] net: cpsw: fix wrong handling of vlan aware mode

	According to TI's datasheet (14.3.2.10.2 Dual Mac Mode)
	ale_vlan_aware bit should be enabled when using
	dual mac. As the comment in this code section statet
	unaware mode should be enabled in !dual-mac mode.

	This fix was testet for both modes.
---
 drivers/net/ethernet/ti/cpsw.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/ti/cpsw.c b/drivers/net/ethernet/ti/cpsw.c
index 8c790ff..b91a583 100644
--- a/drivers/net/ethernet/ti/cpsw.c
+++ b/drivers/net/ethernet/ti/cpsw.c
@@ -1269,12 +1269,16 @@ static void cpsw_init_host_port(struct cpsw_priv *priv)
 	soft_reset("cpsw", &cpsw->regs->soft_reset);
 	cpsw_ale_start(cpsw->ale);
 
-	/* switch to vlan unaware mode */
-	cpsw_ale_control_set(cpsw->ale, HOST_PORT_NUM, ALE_VLAN_AWARE,
-			     CPSW_ALE_VLAN_AWARE);
-	control_reg = readl(&cpsw->regs->control);
-	control_reg |= CPSW_VLAN_AWARE;
-	writel(control_reg, &cpsw->regs->control);
+	/* switch to vlan aware mode when dual_emmac mode is enabled.
+	 * According to the TI's datasheet vlan aware mode is disabled by default */
+	if (cpsw->data.dual_emac) {
+		cpsw_ale_control_set(cpsw->ale, HOST_PORT_NUM, ALE_VLAN_AWARE,
+				     CPSW_ALE_VLAN_AWARE);
+		control_reg = readl(&cpsw->regs->control);
+		control_reg |= CPSW_VLAN_AWARE;
+		writel(control_reg, &cpsw->regs->control);
+	}
+
 	fifo_mode = (cpsw->data.dual_emac) ? CPSW_FIFO_DUAL_MAC_MODE :
 		     CPSW_FIFO_NORMAL_MODE;
 	writel(fifo_mode, &cpsw->host_port_regs->tx_in_ctl);
-- 
2.7.4

